### ISSUE TITLE
Memory dependencies between aliased resources

### ISSUE THREAD

**Issue by philiptaylor** [KhronosGroup/Vulkan-Docs#306]
https://www.khronos.org/registry/vulkan/specs/1.0/xhtml/vkspec.html#synchronization-buffer-memory-barrier says:

> The buffer memory barrier ... only applies to memory accesses involving a specific range of the specified buffer object. That is, a memory dependency formed from a buffer memory barrier is scoped to the specified range of the buffer.

where "scoped" links to:

> Buffer and image memory barriers apply to the buffer range(s) or image subresource(s) included in the command. For accesses to a byte of a buffer or image subresource of an image to be synchronized between two sets of commands, the byte or image subresource must be included in both the first and second halves of the dependencies described above ...

I don't really understand the intended semantics here, despite it being stated four times.

Does it mean the dependency applies to accesses involving a specific range of bytes in memory, where that range happens to be determined by the memory bound to the specified buffer object? Or does it include only accesses that are performed via a command on that particular `VkBuffer` (perhaps indirectly through a descriptor or a `VkBufferView`), and _not_ accesses to the same bytes through a different `VkBuffer`?

For example, say I have two buffers B1 and B2 that are bound to the same offset and size in device memory, so they are aliased. I want to write to B1 (through a `STORAGE_BUFFER` descriptor in a shader), do the necessary barriers as required in the aliasing section, then read from B2 (through another `STORAGE_BUFFER`). Is it sufficient to have a single `VkBufferMemoryBarrier` with `srcAccessMask=SHADER_WRITE, dstAccessMask=SHADER_READ, buffer=B1`? Or do I have to have two barriers, one with `srcAccessMask=SHADER_WRITE, buffer=B1` and a second one with `dstAccessMask=SHADER_READ, buffer=B2`?

`VkImageMemoryBarrier` has similar language, so the same question applies to any combinations of aliasing buffers and images.

For example, say I have a buffer B1 and an optimal-tiled general-layout image I1 that are bound to the same offset and size in device memory, so they are aliased. I write to I1 as a framebuffer attachment; then do a `VkBufferMemoryBarrier` with `srcAccessMask=COLOR_ATTACHMENT_WRITE, dstAccessMask=SHADER_READ, buffer=B1`; then I read from I1 as a sampled image. Will that work? Or do I have to have a `VkImageMemoryBarrier` with `image=I1` instead?

Or for the opposite example, I write to B1 as a storage buffer; then do a `VkImageMemoryBarrier` with `srcAccessMask=SHADER_WRITE, dstAccessMask=SHADER_READ, image=I1`; then I read from B1 as a storage buffer. Will that work? Or do I have to have a `VkBufferMemoryBarrier` with `buffer=B1` instead?

---

If it's the first option, i.e. all that matters is the range of bytes accessed: That seems pretty awkward in the image/buffer case, since I guess the implementation might not have a clear mapping between the reported size of the image (from `vkGetImageMemoryRequirements`) and the range of bytes that it uses in the hardware - it might really use less (e.g. if it reported lots of padding for alignment reasons) or more (e.g. if there's some hidden compression metadata in different memory). So I don't quite see how that would work sanely.

If it's the second option, i.e. the barrier only applies to accesses through that particular buffer object: I think that makes it surprisingly hard to create barriers on host reads/writes - host accesses happen through a mapped `VkDeviceMemory`, they are not related to any particular `VkBuffer`, so you can't use a `VkBufferMemoryBarrier` to make the host writes available or make things visible to host reads. You can probably still use a global `VkMemoryBarrier`, but that seems overkill and quite non-obvious.

(I care about this since I'm thinking about validation of memory dependencies, and don't know whether I need to track exactly which buffer/image is involved in all the accesses and dependencies, or just the range of bytes.)

**Comment by TobiasHector** [comment]
So this is your "second option". Theoretically every buffer is going to have its own virtual address range, which is what the "buffer range" translates into, and most caches are based on virtual rather than physical accesses. Buffer/ImageMemoryBarriers basically translate the range/subresources into relevant ranges of virtual addresses and flush or invalidate those as appropriate.

So for host access with Buffer/ImageBarriers, the idea is that you'd flush it on the host (via vkFlushMappedMemoryRange or using the COHERENT bit) - which makes the memory available; then an image/memory barrier would just be saying "the host made this available, I want it visible to this range of this resource only for this access/stage". You can't flush *from* the host using a memory barrier in this way.

Hopefully that makes sense...?

**Comment by philiptaylor** [comment]
For non-sparse resources, I thought `VkDeviceMemory` corresponded to contiguous virtual address space, not physical, and two aliased resources would be using the same virtual addresses. It sounds like you're suggesting the opposite?

If `VkDeviceMemory` was physical address space, and every resource had its own unique virtual address range, I think that would require every resource to be page-aligned in `vkBindFooMemory` calls (since the virtual-to-physical mapping is per ~4KB/64KB page), which would be very wasteful and doesn't seem to be what current implementations do. I expect you'd also have terrible fragmentation problems when allocating large `VkDeviceMemory` if they were contiguous physical memory. Also, `vkGetDeviceMemoryCommitment` only makes sense if `VkDeviceMemory` is virtual, given that it treats commitment as a property of the `VkDeviceMemory` and not a property of buffers/images.

If it is virtual, and implementations just translate the buffer/image parameters into a virtual address range for their virtually-indexed cache operations, then two aliased resources would map onto the same virtual range, so I'd expect it to work fine if you write to one alias then flush a different alias (i.e. my first option).

(Sparse images are harder since they must have their own virtual address space per image, but they have their own aliasing rules, and I'd be happy to ignore them for now.)

> So for host access with Buffer/ImageBarriers, the idea is that you'd flush it on the host (via vkFlushMappedMemoryRange or using the COHERENT bit) - which makes the memory available

I believe the spec disagrees:

> Host-visible memory types that advertise the VK_MEMORY_PROPERTY_HOST_COHERENT_BIT property still require memory barriers between host and device in order to be coherent [...]
> For host writes to be seen by subsequent command buffer operations, a pipeline barrier from a source of VK_ACCESS_HOST_WRITE_BIT and VK_PIPELINE_STAGE_HOST_BIT to a destination of the relevant device pipeline stages and access types must be performed.

`vkFlushMappedMemoryRange` only flushes the CPU caches, you still need HOST_WRITE/READ memory barriers to do any GPU-side operations. (And then you have the problem of which buffers/images to specify in those host memory barriers, since the host didn't use any buffers/images.)

(I can imagine a hypothetical implementation which decides that VRAM is really slow so it wants to keep everything in a huge L2 cache as long as possible, so "make available" (srcAccessMask) just means a previous write must have reached L2 (but doesn't need to be flushed to VRAM yet) and "make visible" (dstAccessMask) means a subsequent read/write must fetch the latest value from L2. On that implementation, if host writes are not coherent with L2 and they only end up in VRAM, then a srcAccessMask=HOST_WRITE would make the writes available by invalidating the L2 cache (so anything else fetching via L2 will see what the host wrote into VRAM). The memory dependencies would be necessary in that case.)

**Comment by TobiasHector** [comment]
> For non-sparse resources, I thought VkDeviceMemory corresponded to contiguous virtual address space, not physical, and two aliased resources would be using the same virtual addresses. It sounds like you're suggesting the opposite?

I complicated things by talking about virtual addresses - but no that's not the case. It *might* be the case, but it depends on the implementation. I was really just trying to build a mental model of the intent behind the spec - this is one of the reasons we don't call out virtual vs. physical addresses in the Vulkan spec itself :)

> If VkDeviceMemory was physical address space, and every resource had its own unique virtual address range, I think that would require every resource to be page-aligned in vkBindFooMemory calls (since the virtual-to-physical mapping is per ~4KB/64KB page), which would be very wasteful and doesn't seem to be what current implementations do. I expect you'd also have terrible fragmentation problems when allocating large VkDeviceMemory if they were contiguous physical memory. Also, vkGetDeviceMemoryCommitment only makes sense if VkDeviceMemory is virtual, given that it treats commitment as a property of the VkDeviceMemory and not a property of buffers/images.

Yea I definitely over-complicated things. Don't get too hung up on virtual vs. physical for anything other than the specific case I'm calling out - it's just a mental model.

> If it is virtual, and implementations just translate the buffer/image parameters into a virtual address range for their virtually-indexed cache operations, then two aliased resources would map onto the same virtual range, so I'd expect it to work fine if you write to one alias then flush a different alias (i.e. my first option).

As per my first comment - it *might* do that. But the spec doesn't guarantee that because it doesn't guarantee that a VkDeviceMemory object is associated with a specific virtual address.

> I believe the spec disagrees:

> > Host-visible memory types that advertise the VK_MEMORY_PROPERTY_HOST_COHERENT_BIT property still require memory barriers between host and device in order to be coherent [...]
For host writes to be seen by subsequent command buffer operations, a pipeline barrier from a source of VK_ACCESS_HOST_WRITE_BIT and VK_PIPELINE_STAGE_HOST_BIT to a destination of the relevant device pipeline stages and access types must be performed.

Ok I may have over-used the term "available". It makes it available to be made... available? Yea that's terrible terminology, but the host needs to guarantee it's somewhere that the barrier knows where the heck to get it from.

I think I see the problem now - you want to make available (Vulkan, Device available!) a specific memory range to the device. Right - so no, AFAIK you can't do that in any way that it would be guaranteed to work.

**Comment by jeffbolznv** [comment]
> Does it mean the dependency applies to accesses involving a specific range of bytes in memory, where that range happens to be determined by the memory bound to the specified buffer object? Or does it include only accesses that are performed via a command on that particular VkBuffer (perhaps indirectly through a descriptor or a VkBufferView), and not accesses to the same bytes through a different VkBuffer?

I don't think we ever discussed this subtlety in the group, but I am leaning toward the first option. There's some spec language to support that option where the meaning of each parameter is defined (emphasis mine):

```
  * pname:buffer is a handle to the buffer whose **backing memory** is affected
    by the barrier.
  * pname:offset is an offset in bytes into the **backing memory** for
    pname:buffer; this is relative to the base offset as bound to the buffer
    (see flink:vkBindBufferMemory).
  * pname:size is a size in bytes of the affected area of **backing memory** for
    pname:buffer, or ename:VK_WHOLE_SIZE to use the range from pname:offset
    to the end of the buffer.
```

Images similarly talk about "backing memory". So IMO the barrier applies to the range(s) of backing memory (i.e the VkDeviceMemory object(s)) that are bound to the buffer range or image subresource. Which bytes are affected is fairly clear for a "host-accessible resource" (see the "Memory Aliasing" section), and for a non-host-accessible resource you can't meaningfully synchronize with the host so there's no need to know which bytes are affected (and for device->device synchronization there is no problem if you are not aliasing, and if you are aliasing then apply the rules from the Memory Aliasing section).

In an ideal world it would be nice to refactor the definition of "host-accessible" to not be in the Memory Aliasing section so we could more cleanly leverage its definition for buffer/image memory barriers. But I'd settle for just clarifying this in the paragraph that "scoped" links to (synchronization-execution-and-memory-dependencies-types).

(FWIW, I don't think discussing what is virtual and what is physical is necessary to resolve this. We tried (and mostly succeeded) not to use those terms in normative spec language).

**Comment by oddhack** [comment]
@TobiasHector did you remove the "Resolving Inside Khronos" label because you don't think we need to take any action on this, or was it a typo, or something else?

**Comment by TobiasHector** [comment]
@oddhack If I did it deliberately it would've been because I didn't think a spec change was necessary, but I don't remember doing that, and it seems Jeff thinks an edit might be desirable so.... let's call it an accident?

**Comment by oddhack** [comment]
Thanks. It helps me keep the telecon agenda wiki organized with less overall grungework.

**Comment by TobiasHector** [comment]
@jeffbolznv we may need to pull this discussion to an internal bug at some point, but as we've been going through the sync chapter edits, I don't think the real answer quite matches either option - there's a subtlty being missed.

So when you perform a resource-specific barrier, here's what I now *think* is supposed to happen:

* Writes to the **particular buffer/subresource range** in the memory barrier are made **available**. Writes via other resources are not.
 * Those writes become available to the **backing memory** (i.e. available for the host or any other aliasing resource).
* Those writes are made **visible** only to future accesses of the **particular buffer/subresource range**.

* **vkInvalidateMappedMemoryRanges** can be then called to make available writes **visible** to the host.
 * If the memory has VK_MEMORY_PROPERTY_HOST_COHERENT_BIT set, then it's automatically made visible once it's made available (i.e. don't call vkInvalidateMappedMemoryRanges).

There's definitely some language around Fences which I carried over in my last batch of sync chapter edits that seems to contradict this last part - but I think it's more of an oversight with regards the interaction with COHERENT than anything else - so quite easy to fix.

Jeff, does that sound accurate to you? I think we're getting closer to making this obvious in the spec.

**Comment by jeffbolznv** [comment]
I'd be happy with that interpretation/definition.

**Comment by TobiasHector** [comment]
Excellent. This might be why we disagreed before!

I think the spec sort of says this now, but I'll keep the bug open to make sure it's more obvious in the spec once all the other edits are done.

**Comment by TobiasHector** [comment]
The next weekly update should at least make this *slightly less* ambiguous, but I've got a specific fix in mind for this which should come shortly after.

**Comment by oddhack** [comment]
This should be partly addressed in the 1.0.35 spec update, as part of the synchronization language rewrite.

**Comment by oddhack** [comment]
This should be fixed in the 1.0.39 spec update.

**Comment by plasmacel** [comment]
So what is the resolution of this? Is a single memory barrier enough for the memory dependency between aliased resources?
